### PR TITLE
feat(nodejs): parse licenses in yarn projects

### PR DESCRIPTION
**Pull request by nikpivkin**
## Description

At the moment, dependency licenses in projects with the `yarn` package manager are not parsed. This pr adds license scanning for packages from the cache folders `.yarn` (for yarn 2+) and `node_modules` (yarn classic). The license field is also checked, because it can refer to a file and then it is necessary to classify the license. (https://github.com/npm/npm/issues/8795#issuecomment-119760485, https://docs.npmjs.com/cli/v9/configuring-npm/package-json#license)

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/4811

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 1]
```diff
@@ -177,3 +210,148 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
...
+	}
+	// The file path to package.json - */node_modules/<package_name>/package.json
+	// or */node_modules/@<scope_name>/<package_name>/package.json
+	return len(dirs) > 1 && dirs[nodeModulesIdx] == "node_modules" && fileName == types.NpmPkg
+}
+
+func splitPath(path string) (dirs []string, fileName string) {
+	dir, fileName := filepath.Split(path)
+	dir = strings.TrimRight(dir, "/")
+	dirs = strings.Split(dir, "/")
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:250]:
```suggestion
	fileName := filepath.Base(filePath)
	dirs := strings.Split(path.Dir(filePath), "/")
```
wdyt?

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 2]
```diff
@@ -177,3 +210,148 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
...
+			}
+			pkgFile, err := f.Open()
+			if err != nil {
+				return xerrors.Errorf("file open error: %w", err)
+			}
+			defer pkgFile.Close()
+			pkg, err := a.packageJsonParser.Parse(pkgFile)
+			if err != nil {
+				return xerrors.Errorf("unable to parse %q: %w", path, err)
+			}
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:345]: You are using `defer` in a loop. Can we move this logic into a separate function to close each file?

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 3]
```diff
@@ -177,3 +210,148 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
...
+	nodeModulesPath := path.Join(dir, "node_modules")
+
+	if _, err := fs.Stat(fsys, nodeModulesPath); errors.Is(err, fs.ErrNotExist) {
+		// try to find for yarn v2+
+		return a.traverseYarnModernPkgs(fsys, dir, fn)
+	} else if err != nil {
+		return xerrors.Errorf("unable to parse %q: %w", nodeModulesPath, err)
+	}
+
+	return a.traverseYarnClassicPkgs(fsys, nodeModulesPath, fn)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:228]: I think about 1 case when `node_modules` and `.yarn` folders exist.
E.g. User updated project to yarn v2.
Does `node_modules` folder still exist or will `yarn` remove this folder?

@nikpivkin wdyt? Perhaps we need to check `.yarn` folder first (as latest implementation) or(and) add debug message that 2 folder were found.

File: pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		{
+			"name": "Jordan Harband",
+			"email": "ljharb@gmail.com",
+			"url": "http://ljharb.codes"
+		}
+	],
+	"description": "Is this JS value callable? Works with Functions and GeneratorFunctions, despite ES6 @@toStringTag.",
+	"license": "MIT",
+	"main": "index.js",
+	"scripts": {
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json:22]: I think we can remove the fields we don't need.
as for a `happy` test:
https://github.com/aquasecurity/trivy/blob/main/pkg/fanal/analyzer/language/nodejs/yarn/testdata/happy/package.json

**Review by DmitriyLewen — COMMENTED** [review]
Nice work!
I left few small comments.
@nikpivkin take a look, when you have time, please.

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 3 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:228]: @DmitriyLewen then the user should delete `node_modules`)
In fact, if the yarn [nodeLinker](https://yarnpkg.com/configuration/yarnrc#nodeLinker) is set as `node_modules`, then the packages will be installed in` node_modules`, but the folder `.yarn` can also exist, because it stores not only packages.

File: pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json (hunk 4 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json:22]: Yes, you're right

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 2 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go]: Great catch! Done.

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 1 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go]: Done.

File: pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json (hunk 4 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/testdata/yarn-classic-licenses/node_modules/is-callable/package.json]: done

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:228]: Thanks for information.

I also played with `yarn`:
After updating `yarn` and adding/updating package -> `yarn` moves packages from `node_modules` to `.yarn` folder.
Some example:
```
➜  docker run --rm -it node@sha256:2d5e8a8a51bc341fd5f2eed6d91455c3a3d147e91a14298fc564b5dc519c1666 sh

/ # mkdir test && cd "$_"

/test # yarn set version 1.22.19
Resolving 1.22.19 to a url...
Downloading https://github.com/yarnpkg/yarn/releases/download/v1.22.19/yarn-1.22.19.js...
Saving it into /test/.yarn/releases/yarn-1.22.19.cjs...
Updating /test/.yarnrc...
Done!

/test # yarn add is-callable@1.2.7
...

/test # ls -hl node_modules/  
drwxr-xr-x    4 root     root        4.0K Jun 29 04:40 is-callable

/test # ls -hl ./.yarn
drwxr-xr-x    2 root     root        4.0K Jun 29 04:40 releases

/test # yarn set version 3.4.1
...

/test # yarn add is-odd@3.0.1
...

/test # ls -hl node_modules/
ls: node_modules/: No such file or directory

/test # ls -hl .yarn
drwxr-xr-x    2 root     root        4.0K Jun 29 04:41 cache
-rw-r--r--    1 root     root        1.5K Jun 29 04:41 install-state.gz
drwxr-xr-x    2 root     root        4.0K Jun 29 04:41 releases

/test # ls -hl .yarn/cache
total 28K    
-rw-r--r--    1 root     root       12.3K Jun 29 04:41 is-callable-npm-1.2.7-808a303e61-61fd57d03b.zip
-rw-r--r--    1 root     root        4.2K Jun 29 04:41 is-number-npm-6.0.0-30881e83e6-f73bfced02.zip
-rw-r--r--    1 root     root        3.6K Jun 29 04:41 is-odd-npm-3.0.1-93c3c3f41b-4e2b20764d.zip

```

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go]: We use new function to only parse `package.json` **from zip file**. The name `parsePackageJson` can be confusing (I mean you might think that this function for any `package.json`).
What if we use `parsePackageJsonFromZip` name?

Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go]: done

**Review by DmitriyLewen — COMMENTED** [review]
I have 1 question:
If package is installed in `workspace` - does `yarn` still create copy/link of `package.json` file in `node_mosules`/`.yarn` folder? 
Can you check this case?

Could you also update [nodejs page](https://aquasecurity.github.io/trivy/v0.43/docs/scanner/vulnerability/language/nodejs/) in docs, please.

**Comment by nikpivkin** [comment]
@DmitriyLewen yarn has only 1 `node_modules/.yarn` folder in the root of the project, so all the dependencies of the workspaces are there.

**Comment by DmitriyLewen** [comment]
ok, thanks for the clarification!

**Review by DmitriyLewen — APPROVED** [review]

**Comment by DmitriyLewen** [comment]
@nikpivkin Rebase this PR, please.

**Review by DmitriyLewen — APPROVED** [review]

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 5]
```diff
@@ -242,3 +275,157 @@ func (a yarnAnalyzer) traverseWorkspaces(fsys fs.FS, workspaces []string) ([]pac
...
+	dirs, fileName := splitPath(path)
+	nodeModulesIdx := len(dirs) - 2
+	// the scope starts with "@" https://docs.npmjs.com/cli/v9/using-npm/scope
+	if len(dirs) > 2 && strings.HasPrefix(dirs[len(dirs)-2], "@") {
+		nodeModulesIdx -= 1
+	}
+	// The file path to package.json - */node_modules/<package_name>/package.json
+	// or */node_modules/@<scope_name>/<package_name>/package.json
+	return len(dirs) > 1 && dirs[nodeModulesIdx] == "node_modules" && fileName == types.NpmPkg
+}
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:310]: Do we need this complex logic?

>or in the package tests in node_modules - then not.

Do you mean dummy package.json?

My question is why we cannot simply check if the file name matches `package.json`.

Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:310]: @knqyf263 It can be a dummy file for tests.

Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:310]: Yes, I agree

**Review by AnaisUrlichs — APPROVED** [review]

File: pkg/fanal/analyzer/language/nodejs/license.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+			return xerrors.Errorf("walk error: %w", err)
+		}
+
+		return nil
+	}
+}
+
+// IsLicenseRefToFile The license field can refer to a file
+// https://docs.npmjs.com/cli/v9/configuring-npm/package-json
+func IsLicenseRefToFile(maybeLicense string) (bool, string) {
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license.go:72]: Looks like we can make this function non-exportable.

File: pkg/fanal/analyzer/language/nodejs/license.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				return err
+			}
+
+			// License found
+			if len(findings) > 0 {
+				licenses[pkg.ID] = findings.Names()
+			} else {
+				log.Logger.Debugf(
+					"The license file %q was not found or the license could not be classified", licenseFilePath)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license.go:57]:
```suggestion
				log.Logger.Debugf("The license file %q was not found or the license could not be classified", licenseFilePath)
```

File: pkg/fanal/analyzer/language/nodejs/license.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+		return nil
+	}
+}
+
+// IsLicenseRefToFile The license field can refer to a file
+// https://docs.npmjs.com/cli/v9/configuring-npm/package-json
+func IsLicenseRefToFile(maybeLicense string) (bool, string) {
+	if maybeLicense == "" {
+		return true, "LICENSE"
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license.go:74]: Can you add comment why we use `LICENSE` here?

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 9]
```diff
@@ -78,8 +101,40 @@ func (a yarnAnalyzer) PostAnalyze(_ context.Context, input analyzer.PostAnalysis
...
 
 func (a yarnAnalyzer) Required(filePath string, _ os.FileInfo) bool {
-	fileName := filepath.Base(filePath)
-	return fileName == types.YarnLock || fileName == types.NpmPkg
+	dirs, fileName := splitPath(filePath)
+
+	if fileName == types.YarnLock &&
+		containsAny(filePath, "node_modules", ".yarn/unplugged") {
+		return false
+	}
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:109]: Can you add comments why you use these conditions?

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go [hunk 10]
```diff
@@ -78,8 +101,40 @@ func (a yarnAnalyzer) PostAnalyze(_ context.Context, input analyzer.PostAnalysis
...
+
+	if fileName == types.YarnLock ||
+		fileName == types.NpmPkg ||
+		strings.HasPrefix(strings.ToLower(fileName), "license") {
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:113]: if I understand correctly - any license file name is possible:
https://github.com/aquasecurity/trivy/blob/ba38c06c32db1c091d9bb26fa43a742cecbd94c1/pkg/fanal/analyzer/language/nodejs/license.go#L79-L86
Perhaps we need to save `node_modules`/`.yarn` folder?

File: pkg/fanal/analyzer/language/nodejs/license.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+	dio "github.com/aquasecurity/go-dep-parser/pkg/io"
+	"github.com/aquasecurity/go-dep-parser/pkg/nodejs/packagejson"
+	"github.com/aquasecurity/trivy/pkg/fanal/types"
+	"github.com/aquasecurity/trivy/pkg/licensing"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/utils/fsutils"
+)
+
+func ParseLicenses(
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license.go:19]: `npm` also finds licenses in `node_modules/<package-name>/package.json` file.
Maybe we can use this function? or will these changes be in another PR?

npm also finds licenses in the node_modules/<package-name>/package.json file.
Maybe we can use this feature? Or will these changes be in another PR?

File: pkg/fanal/analyzer/language/nodejs/path.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
+package nodejs
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/path.go:1]: what if use `utils` file name for this file?

**Review by DmitriyLewen — COMMENTED** [review]
@nikpivkin left some comments

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 10 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:113]: The cache folder can be large.

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 11 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license.go:19]: How about making these changes in another PR?

File: docs/docs/scanner/vulnerability/language/nodejs.md [hunk 13]
```diff
@@ -33,7 +33,8 @@ By default, Trivy doesn't report development dependencies. Use the `--include-de
 
 ### Yarn
 Trivy parses `yarn.lock`, which doesn't contain information about development dependencies.
-To exclude devDependencies, `package.json` also needs to be present next to `yarn.lock`.
+To exclude devDependencies, `package.json` also needs to be present next to `yarn.lock`. 
+Trivy analyzes `.yarn` (Yarn 2+) or `node_modules` (Yarn Classic) for licenses.
```
[the remark this answers is not in the document]
Comment (nikpivkin) [reply docs/docs/scanner/vulnerability/language/nodejs.md]: done

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 6 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license.go]: done

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 7 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license.go]: done

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 8 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license.go:73]: done

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 9 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go]: done

File: pkg/fanal/analyzer/language/nodejs/path.go (hunk 12 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/path.go]: done

File: pkg/fanal/analyzer/language/nodejs/yarn/yarn.go (hunk 10 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/nodejs/yarn/yarn.go:114]: I assumed that...
Okay. I think in most cases license file name contains `license`. 
Let's wait for feedback from users.
If users say that there are files with other names - we will return to this question.

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 11 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/nodejs/license.go:19]: Okay. I created #4964 for this task.

File: integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    "keywords": [
+      "jquery",
+      "javascript",
+      "browser",
+      "library"
+    ],
+    "bugs": {
+      "url": "https://github.com/jquery/jquery/issues"
+    },
+    "license": "MIT",
```
Comment (DmitriyLewen) [inline integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json:25]: Your golden file doesn't contain license.

File: pkg/fanal/analyzer/language/nodejs/license.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
+package nodejs
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license.go:1]: This is second time I am thinking of moving this file to `nodejs/license` folder, but I'm not sure about it.
@knqyf263 wdyt?

File: pkg/fanal/analyzer/language/nodejs/license_test.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+	tests := []struct {
+		name    string
+		dir     string
+		want    map[string][]string
+		wantErr string
+	}{
+		{
+			name: "happy",
+			dir:  filepath.Join("testdata", "happy"),
+			want: map[string][]string{
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license_test.go:26]: Can you add a test case with a missing license?
I am referring to case where license field in `package.json` is missing and `LICENSE` file doesn't exist.

File: integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json (hunk 14 above)
Comment (nikpivkin) [reply integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json:25]: Good trick! I forgot to move this file after renaming fs to repo.

File: pkg/fanal/analyzer/language/nodejs/license_test.go (hunk 16 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license_test.go:26]: done

File: integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json (hunk 14 above)
Comment (nikpivkin) [reply integration/testdata/fixtures/fs/yarn/node_modules/jquery/package.json]: done

File: pkg/fanal/analyzer/language/nodejs/license.go (hunk 15 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license.go:1]: sounds good, I'll wait for @knqyf263  answer

Comment (knqyf263) [reply pkg/fanal/analyzer/language/nodejs/license.go:1]: Sounds good

File: pkg/fanal/analyzer/language/nodejs/license/license.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+			}
+
+			ok, licenseFileName := isLicenseRefToFile(pkg.License)
+
+			if !ok {
+				licenses[pkg.ID] = []string{pkg.License}
+				return nil
+			}
+
+			log.Logger.Debugf("Licenses are missing in %q, an attempt to find them in the LICENSE file", pkgJsonPath)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/nodejs/license/license.go:44]: I understood that it can confused - users can think that we only parse `LICENSE` file.
```suggestion
			log.Logger.Debugf("License names are missing in %q, an attempt to find them in the%q file", pkgJsonPath, licenseFileName)
```

**Review by DmitriyLewen — COMMENTED** [review]
LGTM.
@nikpivkin I left 1 comment about debug message. Can you take a look?

File: pkg/fanal/analyzer/language/nodejs/license/license.go (hunk 17 above)
Comment (nikpivkin) [reply pkg/fanal/analyzer/language/nodejs/license/license.go]: fixed

**Review by DmitriyLewen — APPROVED** [review]
LGTM

**Review by knqyf263 — APPROVED** [review]
